gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java

```diff
@@ -0,0 +1,257 @@
+   * contents per content id per reference </h2>
+   *
+   * <p>Walk through each reference(both live and dead) distributively (one spark task for each
+   * reference).
+   *
+   * <p>For each commit in the reference (use DETACHED reference to fetch commits from dead
+   * reference) check it against bloom filter to decide whether its contents in put operation are
+   * globally expired or not. If globally expired, Add the contents to the expired output for this
+   * content id for this reference.
+   *
+   * <p>Overall the contents after or equal to cutoff time and the contents that are mapped to
+   * commit head of live keys at the time of cutoff timestamp will be retained.
+   *
+   * @param session spark session for distributed computation
+   * @return {@link IdentifiedResult} object having expired contents per content id.
+   */
+  public IdentifiedResult identifyExpiredContents(SparkSession session) {
+    try (NessieApiV1 api = GCUtil.getApi(gcParams.getNessieClientConfigs())) {
+      DistributedIdentifyContents distributedIdentifyContents =
+          new DistributedIdentifyContents(session, gcParams);
+      // key is refName:hash and value is its dropped/assigned time.
+      Map<String, Instant> droppedRefTimeMap = new HashMap<>();
+      List<Reference> allRefs = getAllReferences(api, droppedRefTimeMap);
+      long totalCommitsInDefaultReference = getTotalCommitsInDefaultReference(api);
+      // Identify the live contents and return the bloom filter per content-id
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap =
+          distributedIdentifyContents.getLiveContentsBloomFilters(
+              allRefs, totalCommitsInDefaultReference, droppedRefTimeMap);
+      // Identify the expired contents
+      return distributedIdentifyContents.getIdentifiedResults(liveContentsBloomFilterMap, allRefs);
+    }
+  }
+
+  private List<Reference> getAllReferences(
+      NessieApiV1 api, Map<String, Instant> droppedRefTimeMap) {
+    List<Reference> references = api.getAllReferences().get().getReferences();
+    List<Reference> deadReferences = collectDeadReferences(api, droppedRefTimeMap);
+    List<Reference> allRefs = new ArrayList<>(references);
+    if (deadReferences.size() > 0) {
+      allRefs.addAll(deadReferences);
+    }
+    return allRefs;
+  }
+
+  private long getTotalCommitsInDefaultReference(NessieApiV1 api) {
+    ReferenceMetadata defaultRefMetadata = null;
+    try {
+      defaultRefMetadata =
+          api.getReference()
+              .refName(api.getDefaultBranch().getName())
+              .fetch(FetchOption.ALL)
+              .get()
+              .getMetadata();
+    } catch (NessieNotFoundException ex) {
+      // Ignore the failure and use default value.
+    }
+    return (defaultRefMetadata != null && defaultRefMetadata.getNumTotalCommits() != null)
+        ? defaultRefMetadata.getNumTotalCommits()
+        : 0;
+  }
+
+  private static void withConfig(
```

Comment:
Everything in here should be in GCParams. Immutables has built-in support for validation - so please use it.

With that, all the stuff in GCConfigCOnstants can go into GCParams